sdk/go/pulumi/templates/types_builtins.go.template

```diff
@@ -134,20 +134,34 @@ func (o {{.Name}}Output) To{{.Name}}PtrOutputWithContext(ctx context.Context) {{
 {{end}}
 {{end}}
 {{if .DefineElem}}
+// Elem dereferences the pointer value or returns the zero value of the approporiate type if the pointer is nil.
 func (o {{.Name}}Output) Elem() {{.ElemReturnType}}Output {
 	return o.ApplyT(func (v {{.ElementType}}) {{.ElemElementType}} {
-		return *v
+		var ret {{.ElemElementType}}
+		if v != nil {
+			ret = *v
+		}
+		return ret
 	}).({{.ElemReturnType}}Output)
 }
 {{end}}
 {{if .DefineIndex}}
+// Index looks up the i'th element of the array if it is in bounds or returns the zero value of the appropriate
+// type if the index is out of bounds.
 func (o {{.Name}}Output) Index(i IntInput) {{.IndexReturnType}}Output {
 	return All(o, i).ApplyT(func(vs []interface{}) {{.IndexElementType}} {
-		return vs[0].({{.ElementType}})[vs[1].(int)]
+		arr := vs[0].({{.ElementType}})
+		idx := vs[1].(int)
+		var ret {{.IndexElementType}}
+		if idx >= 0 && idx < len(arr) {
```

Comment:
```suggestion
		var ret {{.IndexElementType}}
		if idx := vs[1].(int); idx >= 0 && idx < len(arr) {
```

Reply:
I honestly slightly prefer the original - so will keep it as is for now.